lib/select-builder.js

```diff
@@ -92,232 +77,162 @@ class SelectBuilder extends QueryBuilder {
+    this.operations.avg.push({ field: this.escapeKey(field) });
     return this;
   }
 
   min(field) {
-    this.operations.get('min').push({ field: this.escapeKey(field) });
+    this.operations.min.push({ field: this.escapeKey(field) });
     return this;
   }
 
   max(field) {
-    this.operations.get('max').push({ field: this.escapeKey(field) });
+    this.operations.max.push({ field: this.escapeKey(field) });
     return this;
   }
 
   sum(field) {
-    this.operations.get('sum').push({ field: this.escapeKey(field) });
+    this.operations.sum.push({ field: this.escapeKey(field) });
     return this;
   }
 
-  processSelect(query, clauses) {
-    if (clauses.size > 0) {
-      return query + ' ' + iter(clauses).reduce((acc, id) => acc + ', ' + id);
-    }
-    return query + ' *';
-  }
-
-  processOperations(query, operations, functionHandlers) {
-    for (const fn of Object.keys(functionHandlers)) {
-      const ops = operations.get(fn);
+  buildFields() {
+    const fields = [...this.operations.fields];
+    for (const fn of sqlFunctions) {
+      const ops = this.operations[fn];
       if (ops.length === 0) continue;
-      else if (query.endsWith(' *')) query = query.slice(0, -2);
-      else query += ',';
-      const handler = functionHandlers[fn];
-      query += ops
-        // eslint-disable-next-line no-loop-func
-        .reduce((acc, op) => acc + handler(op, query) + ',', ' ')
-        .slice(0, -1);
+      for (const { field } of ops) fields.push(`${fn}(${field})`);
     }
-    return query;
-  }
-
-  processWhere(query, clauses) {
-    // TODO(lundibundi): support braces
-    query += ' WHERE';
-    for (let i = 0; i < clauses.length; ++i) {
-      const clause = clauses[i];
-      if (i !== 0) {
-        if (clause.or) query += ' OR';
-        else query += ' AND';
-      }
-      if (clause.mod) query += ` ${clause.mod}`;
-      if (clause.key) query += ` ${clause.key}`;
-      query +=
-        ` ${clause.cond} ` +
-        makeParamValue(clause.cond, clause.value, this.params);
+    return fields.length > 0 ? fields.join(', ') : '*';
+  }
+
+  buildWhere() {
+    const { where } = this.operations;
+    const conditions = [];
+    for (const { key, value, op, mod, or } of where) {
+      if (conditions.length > 0) conditions.push(or ? 'OR' : 'AND');
+      if (mod) conditions.push(mod);
+      if (key) conditions.push(key);
+      conditions.push(op);
+      const par = makeParamValue(op, value, this.params);
+      conditions.push(par);
     }
-    return query;
+    return conditions.join(' ');
   }
 
-  processOrder(query, clauses) {
-    const it = iter(clauses);
-    const firstClause = it.next().value;
-    query += ` ORDER BY ${firstClause.field} ${firstClause.dir}`;
-    for (const order of it) {
-      query += `, ${order.field} ${order.dir}`;
+  buildOrder() {
+    const fields = [];
+    for (const { field, desc } of this.operations.order) {
```

Comment:
Ditto.

Reply:
We can't use `Set` here, `ORDER BY` should have certain field order